projectq/meta/_util.py

```diff
@@ -34,6 +34,12 @@ def insert_engine(prev_engine, engine_to_insert):
     engine_to_insert.next_engine = prev_engine.next_engine
     prev_engine.next_engine = engine_to_insert
 
+    if prev_engine.main_engine is not None:
+        prev_engine.main_engine.n_engines += 1
+
+        if prev_engine.main_engine.n_engines > prev_engine.main_engine.n_engines_max:
+            raise RuntimeError('Too many compiler engines added to the MainEngine!')
+
```

Comment:
Optional: Shouldn't this check happen before all the assignments?
It throws a runtime error which will not be caught anywhere currently, so it doesn't matter too much.